---
title: "Lung rejection scRNA-seq data preprocessing"
author: "Kent Riemondy RBI"
date: "`R Sys.Date()`"
params:
  ncores: 4
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(cache.lazy = FALSE)
```


```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source(here::here("R/utils.R"))
library(future)

plan("multicore", workers = 7)
options(future.globals.maxSize = 2 * 1024 ^ 3)

proj_dir <- here()
data_dir <- file.path(proj_dir, "data", "cellranger", "results")
doc_dir <- file.path(proj_dir, "docs")

fig_dir <- "figs_3"
mkrs_dir <- "markers_3"
tbls_dir <- "tables_3"
object_dir <- "objects_3"
walk(c(fig_dir, mkrs_dir, tbls_dir, object_dir), 
     dir.create, showWarnings = F)
```

# Experiment Summary

```{r get_data}

samples_1 <- paste0("Eickelberg_", 
                  c("B6B6_1",
                    "B6B6_2",
                    "B6B6_3",
                    "HLAB6_2", 
                    "HLAB6_3"))

samples_2 <- c(
  "1_B_65_11152018",
  "1_B6_B6_11132018",
  "2_HLA_5_11152018")

samples_3 <- c(
  "NaiveB6",
  "B6_B6_171501",
  "HLA_B6_163107",
  "HLA_B6_18467"
)

samples <- c(samples_1, samples_2, samples_3)

sample_paths <- file.path(data_dir,
                          samples,
                          "outs", 
                          "filtered_feature_bc_matrix")

names(sample_paths) <- c(str_remove(samples_1, "Eickelberg_"),
                         "B6B6_5",
                         "B6B6_4",
                         "HLAB6_5",
                         "Naive_B6",
                         "B6B6_6",
                         "HLAB6_6",
                         "HLAB6_7")
```

```{r, eval = FALSE}
mat <- Read10X(sample_paths)
```

## General QC for library prep  {.tabset}

```{r }
metrics_paths <- file.path(data_dir,
                          samples,
                          "outs", 
                          "metrics_summary.csv")

names(metrics_paths) <- str_remove(samples, "Eickelberg_")

mapping_dat <- map_dfr(metrics_paths, read_csv, .id = "sample")

clean_up_metadata <- function(metrics_summary) {
  metrics_summary <- mutate_all(metrics_summary, str_remove, "%$")
  metrics_summary <- mutate_at(metrics_summary, .vars= 2:ncol(metrics_summary), as.numeric)
  metrics_summary
}

mapping_dat <- clean_up_metadata(mapping_dat)

metrics <- colnames(mapping_dat)[2:ncol(mapping_dat)]

mapping_dat <- mapping_dat %>% 
  gather(metric, value, -sample) %>%
  mutate(sample = factor(sample, levels = unique(sample)))

p <- map(metrics, 
    function(x) {
    filter(mapping_dat, metric == x) %>% 
          ggplot(aes(sample, value)) +
            geom_point(aes(color = sample)) +
        scale_color_brewer(palette = "Paired") + 
        labs(y = x, 
             x = "") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
}) 
```

```{r, results ='asis'}

for(i in seq_along(p)){
  cat('\n### ', metrics[i], '\n')
  print(p[i])
  cat('\n')
}
```


## Preprocessing {.tabset}


## Rerun preprocessing scaling, PCA, UMAP, etc. 
    
```{r create_seurat, message = F, results = 'hide', warning = F, eval = FALSE}

so <- CreateSeuratObject(
  mat,
  min.cells = 3,
  min.features = 200,
  names.delim = "_",
  names.field = c(1,2)
)

so <- PercentageFeatureSet(so, 
                             pattern = "^mt-", 
                             col.name = "percent.mt")

so@meta.data <- so@meta.data %>% 
  tibble::rownames_to_column("cell") %>% 
  mutate(condition = str_split(orig.ident, "_", simplify = TRUE) %>% .[, 1] ) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell")


dir.create(object_dir, showWarnings = FALSE)

saveRDS(so, file.path(object_dir, "unfiltered.rds"))
```

```{r}
so <- readRDS(file.path(object_dir, "unfiltered.rds"))

```
### Percent Mitochondrial UMIs 

```{r}
plot_violin(so@meta.data, 
            "orig.ident",
            "percent.mt",
            .fill = "condition") +
  labs(x = "", y = "Percent UMIs from Mitochondria")
```


### # of genes detected 

```{r}
plot_violin(so@meta.data,
            "orig.ident", 
            "nFeature_RNA",
            .fill = "condition") +
  labs(x = "", y = "# of genes per cell")
```


### # of UMIs detected

```{r}
plot_violin(so@meta.data, 
            "orig.ident",
            "nCount_RNA", 
            .fill = "condition") +
  labs(x = "", y = "# of UMIs") 
```

### Table of mitochondrial proportions per sample

```{r}
so@meta.data %>% 
  group_by(orig.ident) %>% 
  summarize(median_percent_mito = median(percent.mt), 
            mean_percent_mito = mean(percent.mt)) %>% 
  arrange(desc(median_percent_mito))

```


## Relationship between UMIs and % mitochondria {.tabset}

### All samples

```{r, fig.width = 24, fig.height = 18}
sample_names <- as.character(unique(so@meta.data$orig.ident))
per_sample <- map(sample_names, ~filter(so@meta.data, 
                                        orig.ident == .x))
p <- list()
for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  p[[i]] <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = condition)) +
        scale_color_manual(values = .col)
}

plot_grid(plotlist= p, nrow = 4, ncol = 3)
```

```{r, results ='asis'}

for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  cat('\n### ', sample_names[i], '\n')
  p <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = condition)) +
        scale_color_manual(values = .col)
  print(p)
  cat('\n')
}
```


## Filter cells and samples.

No filtering applied at this point.

```{r, eval = FALSE}
so <- readRDS(file.path(object_dir, "unfiltered.rds"))
```


Annotate samples and assign colors to each sample. 

```{r, eval = FALSE}
annots <- so@meta.data %>% 
  select(orig.ident, condition) %>% 
  unique() %>%
  arrange(condition)

subgroup_order <- c(
  "B6B6",
  "HLAB6",
  "Naive"
)

annots <- mutate(annots, condition = factor(condition, levels = subgroup_order))

sample_order <- annots$orig.ident 

so@meta.data$orig.ident <- factor(so@meta.data$orig.ident, levels = sample_order)
so@meta.data$condition <- factor(so@meta.data$condition, levels = subgroup_order)

```


## Normalize and embed into 2D with UMAP

```{r, eval = FALSE}

so <- NormalizeData(so)

so <- FindVariableFeatures(
  so,
  selection.method = "vst",
  nfeatures = 2000,
  verbose = FALSE
)

so <- ScaleData(so, verbose = TRUE)

so <- RunPCA(so, 
             npcs = 100, 
             verbose = FALSE)

ElbowPlot(so, ndims = 100)

# make graphs and use graphs for UMAP
so <- FindNeighbors(so, 
                      reduction = "pca", 
                      dims = 1:30, 
                      k.param = 20L)


so <- RunUMAP(so, 
              graph = "RNA_snn", 
              min.dist = 0.3)

res <- c(0.05, 0.075, 0.1, 0.3, 0.5)
so <- FindClusters(so, 
                   resolution = res)

map(str_c("RNA_snn_res.", res), 
    ~plot_umap(so, .x)) %>% 
  plot_grid(plotlist = .,
            nrow = 3, 
            ncol = 2)

```

```{r, eval = FALSE}
so$coarse_clusters <- so$RNA_snn_res.0.1

Idents(so) <- "coarse_clusters"

plot_umap(so, "coarse_clusters")
```




```{r, eval = FALSE}
saveRDS(so, file.path(object_dir, "so.rds"))
```

```{r}
so <- readRDS(file.path(object_dir, "so.rds"))

plot_umap(so, "condition", .col = palette_OkabeIto)
```

```{r}
plot_umap(so, "coarse_clusters", label_text = TRUE, label_col = "black")
```

```{r}
plot_umap(so, "orig.ident")
```

```{r}

plot_features_split(so, "orig.ident", group = "condition")


```

## markers 


### Cluster markers
```{r, eval = FALSE}
so <- readRDS(file.path(object_dir, "so.rds"))
Idents(so) <- "coarse_clusters"
mkrs <- FindAllMarkers(so, only.pos = TRUE)

write_tsv(mkrs, 
          file.path(mkrs_dir, "coarse_cluster_markers_all_data.tsv"))
```



```{r}
so <- readRDS(file.path(object_dir, "so.rds"))
mkrs <- read_tsv(file.path(mkrs_dir, "coarse_cluster_markers_all_data.tsv"))

topx <- mkrs %>% 
  group_by(cluster) %>% 
  top_n(10, avg_logFC) 
```



```{r, fig.width = 20, fig.height= 20}

p <- DotPlot(so, 
        features = unique(topx$gene),
        group.by = "coarse_clusters") +
  coord_flip() +
  labs(x = "Cluster",
       y = "")

p

save_plot(file.path(fig_dir, 
                    "coarse_cluster_marker_dotplot.pdf"), 
          p, base_height = 24, base_asp = 1)
```



## Try to assign cell types using cell atlas data

The [`Tabula Muris`](https://tabula-muris.ds.czbiohub.org/) is a collection of single cell dataset from 20 organs. 

We will use the gene expression profiles from these datasets to try to assign cell types to the identified clusters. The `clustifyr` package will correlate experession in our clusters to the tabula muris and report the most correlated cell type. It's unlikely that these will be 100% correct, because I don't believe that the tabula muris will have all of the cell types in a mouse, and I don't believe hat they have cells from developing tissues. Nevertheless it will give us a good starting point to annotating the cell types. 

```{r, fig.width=8, fig.height=12}
library(clustifyr)
library(ComplexHeatmap)

so <- readRDS(file.path(object_dir, "so.rds"))

tm_average <- readRDS("/Users/kriemo/Projects/sc_repos/vanotterloo/data/tabula_muris/TM_avg_expr.rds")

mdata <- get_metadata(so)

res <- clustify(so@assays$RNA@data, 
                tm_average, 
                query_genes = so@assays$RNA@var.features,
                metadata = mdata, 
                cluster_col = "coarse_clusters", 
                compute_method = "spearman")

hmap <- Heatmap(t(res), 
                viridis::viridis(256), 
                "Spearman\ncorrelation",
                row_title = "Cell types from Tabula Muris",
                column_title = "Clusters")
  
pdf(file.path(fig_dir, "cell_type_heatmap.pdf"), height = 12, width = 12)
draw(hmap)
dev.off()

hmap

```

Reuse existing annotations
```{r, eval = FALSE}
so2 <- readRDS(file.path("objects_2", "so.rds"))

so2_mdata <- get_metadata(so2)
rm(so2)
gc()

so_mdata <- get_metadata(so) %>% 
  select(-(PC_1:UMAP_2))

so2_mdata_simple <- so2_mdata %>% 
  select(cell, cell_types)
combined_mdata <- left_join(so_mdata, so2_mdata_simple, by = "cell")

best_cell_type <- combined_mdata %>% 
  group_by(cell_types, coarse_clusters) %>% 
  summarize(n = n()) %>% 
  group_by(cell_types) %>% 
  summarize(coarse_clusters = nth(coarse_clusters, which(n == max(n))))


so@meta.data <- so_mdata %>% 
  left_join(best_cell_type, by = "coarse_clusters") %>% 
  column_to_rownames("cell")
```


```{r, eval = FALSE}
pclusters <- plot_umap(so, "coarse_clusters")
pcell_types <- plot_umap(so, "cell_types")

plt <- plot_grid(pclusters, pcell_types)
save_plot(file.path(fig_dir,"umap_by_cell_type.pdf"),
          plt, nrow = 1, ncol = 2, base_asp = 1.75)

saveRDS(so, file.path(object_dir, "so.rds"))
```


## Decontamination with SoupX

```{r, eval = FALSE}
library(SoupX)

dataDirs <- file.path(data_dir, samples, "outs")
scl <- load10X(dataDirs)
scl <- inferNonExpressedGenes(scl)

map(str_c("Channel", 1:12),
  ~rownames(scl$channels[[.x]]$nonExpressedGenes)[seq(50)]
) %>% 
  unlist() %>% 
  unique() %>% 
  .[!str_detect(., "^Rp[ls]")] %>% 
  .[!str_detect(., "^mt")]

soup_genes <- list(
  HB = c("Hbb-bs", "Hba-a1", "Hba-a2", "Hbb-bt", "H2-Ab1", "H2-Aa", "H2-Ab1", "H2-Eb1"),
  Club = c("Scgb1a1"),
  Bcell = c("Cd74", "Ighg2c", "Igkc", "Ighm", "Ighg3"),
  AEC2 = c("Sftpc"),
  MAC = c("Apoe", "Lyz2", "Psap" ),
  OTHER = c("Ly6c1", "Ly6a" )
)


for(channel in str_c("Channel", 1:12)){
  message(channel)
  scl <- calculateContaminationFraction(scl, channel, soup_genes)
}

for(channel in str_c("Channel", 1:12)){
  message(channel)
  gg = plotChannelContamination(scl, channel)
  plot(gg)
}

for(channel in str_c("Channel", 1:12)){
  message(channel)
  scl = interpolateCellContamination(scl, channel, useGlobal = FALSE)
}

scl <- strainCells(scl)
scl <- adjustCounts(scl)

rename_cols <- names(sample_paths)

names(rename_cols) <- str_c("Channel", 1:12)

count_matrix <- scl$atoc

channels <- colnames(count_matrix) %>% 
  str_match("Channel[0-9]+") %>%
  .[, 1]

new_ids <- rename_cols[channels]

cell_ids <- str_remove(colnames(count_matrix),
                       "Channel[0-9]+___")

new_ids <- str_c(new_ids, "_", cell_ids) %>% 
  str_remove("-[0-9]+$")

colnames(count_matrix) <- new_ids
```

## renormalize/cluster/embed via seurat

```{r, eval = FALSE}
count_matrix <- count_matrix[rownames(so), colnames(so)]

cell_types <- so$cell_types
so <- CreateSeuratObject(counts = count_matrix , 
                         names.field = c(1, 2))

so$cell_types <- cell_types
so$condition <- str_split(so$orig.ident, "_", simplify = TRUE) %>% .[, 1] 
so <- NormalizeData(so)

so <- FindVariableFeatures(
  so,
  selection.method = "vst",
  nfeatures = 2000,
  verbose = FALSE
)

so <- ScaleData(so, verbose = TRUE)

so <- RunPCA(so, 
             npcs = 100, 
             verbose = FALSE)

ElbowPlot(so, ndims = 100)

# make graphs and use graphs for UMAP
so <- FindNeighbors(so, 
                      reduction = "pca", 
                      dims = 1:30, 
                      k.param = 20L)

so <- RunUMAP(so, 
              graph = "RNA_snn", 
              min.dist = 0.3)

res <- c(0.05, 0.075, 0.1, 0.3, 0.5)
so <- FindClusters(so, 
                   resolution = res)

map(str_c("RNA_snn_res.", res), 
    ~plot_umap(so, .x)) %>% 
  plot_grid(plotlist = .,
            nrow = 3, 
            ncol = 2)

```

```{r, eval = FALSE}
so$coarse_clusters <- so$RNA_snn_res.0.1

Idents(so) <- "coarse_clusters"

plot_umap(so, "coarse_clusters")
plot_umap(so, "cell_types")
plot_umap(so, "condition")
```


```{r, eval = FALSE}
saveRDS(so, file.path(object_dir, "so_souped.rds"))
```

```{r, eval = FALSE}
to_plot <- c(
  "coarse_clusters",
  "cell_types",
  "condition",
  "orig.ident",
  "Mzb1",
  "Cd19"
) 

plt <- map(to_plot, 
    ~plot_umap(so, .x)) %>% 
  plot_grid(plotlist = ., nrow = 3, ncol = 2,
            rel_widths = c(1.2, 2, 1.2, 1.2, 1.2, 1.2))


save_plot(file.path(fig_dir,
                         paste0("umap_summary_soupx.pdf")),
                plt,
          nrow = 3, 
          ncol = 2,
          base_asp = 1.2)
```


## markers 


### Cluster markers
```{r, eval = FALSE}
so <- readRDS(file.path(object_dir, "so_souped.rds"))

mkrs <- FindAllMarkers(so, only.pos = TRUE)

write_tsv(mkrs, 
          file.path(mkrs_dir, "coarse_cluster_markers_all_data_soupx.tsv"))
```


## Overlay VDJ data

```{r add_clonotypes, eval = FALSE}

vdj_subdirs <- c(
  "NaiveB6_Bcell",
  "B6_B6_171501_Bcell",
  "HLA_B6_163107_Bcell",
  "HLA_B6_168467_Bcell")

vdj_dirs <- file.path("/Users/kriemo/Projects/sc_repos/eickelberg/data/vdj",
                      vdj_subdirs,
                      "outs")

prefixes <- str_c(names(sample_paths)[9:12], "_")

out <- map2_dfr(vdj_dirs, prefixes, get_clonotypes)

so_tmp <- add_clonotypes(so, 
                      vdj_dirs, 
                      prefixes = prefixes)

so_tmp$frequency <- ifelse(is.na(so_tmp$frequency), 
                           0L,
                           so_tmp$frequency)

so_tmp$proportion <- ifelse(is.na(so_tmp$proportion), 
                           0L,
                           so_tmp$proportion)

so_tmp$clonotype_present <- ifelse(is.na(so_tmp$cdr3s_aa), 
                           "None",
                           "Present")

c_mat <- so_tmp@meta.data %>% 
  tibble::rownames_to_column("cell") %>%
  mutate(clonotype_id = str_c(str_sub(orig.ident, -1), 
                              "-", 
                              raw_clonotype_id)) %>% 
  select(cell, clonotype_id) %>% 
  mutate(present = 1L,
         clonotype_id = ifelse(is.na(clonotype_id), 
                               "no_clonotype", 
                               clonotype_id),
         clonotype_id =  ifelse(clonotype_id == "None", 
                                "no_clonotype", 
                                clonotype_id)) %>% # need to figure out where None came from...
  spread(cell, present, fill = 0L) %>% 
  tibble::column_to_rownames("clonotype_id") %>% 
  as.data.frame() %>% 
  as.matrix()

c_mat <- c_mat[, colnames(so_tmp@assays$RNA@data)]

c_mat <- as(c_mat, "dgCMatrix")

so_tmp@assays[["tcr"]] <- CreateAssayObject(counts = c_mat)
```

```{r, eval = FALSE}
saveRDS(so_tmp, file.path(object_dir, "so_souped_w_vdj.rds"))
```


```{r}

so <- readRDS(file.path("objects_3", "so_souped_w_vdj.rds"))

to_plot <- c(
  "clonotype_present",
  "frequency",
  "proportion"
)

map(to_plot, 
    ~plot_umap(so, .x)) %>% 
  plot_grid(plotlist = ., nrow = 1, ncol =3) %>% 
  save_plot("figs_3/clonotypes.pdf", ., nrow =1 ,ncol =3, base_asp = 1.2)
```