---
title: "Figures"
author: "Kent Riemondy RBI"
date: "2/9/2020"
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source(here::here("R/utils.R"))
proj_dir <- here()
data_dir <- file.path(proj_dir, "data", "cellranger", "results")
doc_dir <- file.path(proj_dir, "docs")

fig_dir <- "figs"
mkrs_dir <- "markers"
tbls_dir <- "tables"
object_dir <- "objects"
walk(c(fig_dir, mkrs_dir, tbls_dir, object_dir), 
     dir.create, showWarnings = F)

qread <- function(file, ...){qs::qread(file, use_alt_rep = FALSE, nthreads = 4, ...)}
```

## All cells analysis

```{r}
all_cell_outdir <- file.path(fig_dir, "all_cells")
dir.create(all_cell_outdir, showWarnings = FALSE)
```

```{r}
so <- qread(file.path("objects", "so_souped.qs"))
```


### UMAP plots {.tabset}


#### Cell type


```{r}
p <- plot_umap(so, 
          feature = "cell_types", 
          sorted = "random",
          pt_alpha = 1, 
          legend_title = "")

p
save_plot(file.path(all_cell_outdir,
                    "umap_by_cell_type.pdf"),
          p,
          base_asp = 1.6)
```


#### Condition

```{r}
p <- plot_umap(so, 
          feature = "condition", 
          sorted = "random",
          .cols = palette_OkabeIto,
          pt_alpha = 0.6, 
          legend_title = "")

p
save_plot(file.path(all_cell_outdir,
                    "umap_by_condition.pdf"),
          p,
          base_asp = 1.6)
```

#### Cell types + Condition

```{r}
p <- plot_umap(so, 
          feature = "cell_types", 
          sorted = "random",
          group = "condition",
          pt_alpha = 0.6, 
          legend_title = "")

p
save_plot(file.path(all_cell_outdir,
                    "umap_by_condition_split.pdf"),
          p, ncol = 2,
          base_asp = 1.5)
```
#### Sample

```{r}
p <- plot_umap(so, 
          feature = "orig.ident", 
          sorted = "random",
          .cols = palette_OkabeIto,
          pt_alpha = 1, 
          legend_title = "")

p
save_plot(file.path(all_cell_outdir,
                    "umap_by_sample.pdf"),
          p,
          base_asp = 1.6)
```

### UMAP plots by gene {.tabset}

```{r, results ='asis'}
to_plot <- c(
  "Tmem100",
  "Cd3g",
  "C1qb",
  "Cd79a",
  "Plac8",
  "S100a8",
  "Gzma",
  "Cd8b1",
  "Pclaf",
  "Emp2",
  "Hbb-bs",
  "Mzb1",
  "Fcer1a",
  "Krt8",
  "Sftpc")

dir.create(file.path(all_cell_outdir, "umap_by_gene"), showWarnings = FALSE)
res <- list()
for(i in seq_along(to_plot)){
  cat('\n#### ', to_plot[i], '\n')
  
  p <- plot_umap(so, to_plot[i])
  
  print(p)
  res[[i]] <- p
  save_plot(file.path(all_cell_outdir,
                      "umap_by_gene",
                    paste0("umap_by_", to_plot[i], ".pdf")),
          p,
          base_asp = 1.3)
  cat('\n')
}
```

#### All markers

```{r, fig.width = 12, fig.height = 16}
p <- plot_grid(plotlist = res, 
          nrow = 4,
          ncol = 4)

p
save_plot(file.path(all_cell_outdir,
                     "umap_all_selected_markers.pdf"),
          nrow = 4, ncol = 4,
          p,
          base_asp = 1.3)
 
```

#### All markers minimal theme

```{r, fig.width = 12, fig.height = 16}
p <- plot_umap(so, to_plot, minimal_theme = TRUE) %>% 
  plot_grid(plotlist = ., 
          nrow = 4,
          ncol = 4)

p
save_plot(file.path(all_cell_outdir,
                     "umap_all_selected_markers_minimal_theme.pdf"),
          nrow = 4, ncol = 4,
          p,
          base_asp = 1.0)
 
```


### Violin plots {.tabset}

```{r, results ='asis'}
to_plot <- c(
  "C1qb",
  "Cd79a",
  "Mzb1",
  "Plac8",
  "S100a8",
  "Tmem100",  
  "Fcer1a",
  "Pclaf",
  "Gzma",
  "Cd8b1",
  "Cd3g",
  "Emp2",
  "Krt8",
  "Sftpc",
  "Hbb-bs")

dir.create(file.path(all_cell_outdir, "vln_by_gene"), showWarnings = FALSE)
res <- list()
for(i in seq_along(to_plot)){
  cat('\n#### ', to_plot[i], '\n')
  
  p <- scbp::plot_violins(so, 
            group = "cell_types", 
            features = to_plot[i]) +
  labs(x = "") + 
  theme(legend.position = "none")
  print(p)
  
  res[[i]] <- p
  save_plot(file.path(all_cell_outdir,
                      "vln_by_gene",
                    paste0("vln_by_", to_plot[i], ".pdf")),
          p,
          base_asp = 1.5)
  cat('\n')
}

```

#### All markers

```{r}
p <- scbp::plot_violins(so, 
            group = "cell_types", 
            features = to_plot) +
  labs(x = "",
       y = "") + 
  theme(legend.position = "none")
p
save_plot(file.path(all_cell_outdir,
                     "vln_all_selected_markers.pdf"),
          p,
          base_asp = 1.0,
          base_height = 14)

```

#### All markers with condition

```{r}
p <- scbp::plot_violins(so, 
            group = "cell_types", 
            features = to_plot,
            split_by = "condition") +
  labs(x = "",
       y = "") 
p
save_plot(file.path(all_cell_outdir,
                     "vln_all_selected_markers_by_condition.pdf"),
          p,
          base_asp = 1.0,
          base_height = 14)

```

### Cell Proportions {.tabset}

#### Condition by cell type

```{r}
p <- plot_cell_proportions(so, "cell_types", "condition") +
  labs(x = "",
       fill = "") 
  
p

save_plot(file.path(all_cell_outdir, 
                    "cell_proportion_cell_types_by_condition.pdf"),
          p,
          base_asp = 1.2,
          base_height = 5)
```


#### Condition by sample

```{r}
p <- plot_cell_proportions(so, "cell_types", "orig.ident") +
  labs(x = "",
       fill = "")
p

save_plot(file.path(all_cell_outdir, 
                    "cell_proportion_cell_types_by_sample.pdf"),
          p,
          base_asp = 1.2,
          base_height = 5)
```

## Better represent samples

```{r}
x <- get_cell_count_matrix(so, "cell_types", "orig.ident") 
x <-  t(t(x) / colSums(x)) 
plt_dat <- x %>% 
  as.data.frame() %>% 
  rownames_to_column("cell_type") %>% 
  pivot_longer(cols = -cell_type,
               names_to = "sample",
               values_to = "count") %>% 
  mutate(count = 100 * count) %>% 
  separate(sample, c("genotype", "rep"), sep = "_") 

p <- ggplot(plt_dat, aes(genotype, count)) +
  geom_boxplot(aes(fill =genotype), outlier.shape = NA) +
  geom_jitter(aes(group = genotype)) +
  facet_wrap(~cell_type, scales = "free", strip.position = "top", nrow = 3) +
  scale_fill_brewer(palette = "Paired") + 
  labs(x = "",
       y = "% of cells") + 
  theme(strip.placement = "outside",
        legend.position = "none")
p
save_plot(file.path(all_cell_outdir, 
                    "cell_proportion_cell_types_by_sample_all_samples.pdf"),
          p,
          base_asp = 1.5,
          base_height = 6)

cell_type_stats <- calc_da(so, "cell_types", "orig.ident", "condition")

write_tsv(cell_type_stats, file.path(all_cell_outdir, 
                    "cell_proportion_cell_types_by_sample_all_samples_stats.tsv"))
```

### Heatmaps

#### All data 
```{r, fig.width=12, fig.height=8}
mkrs <- read_tsv("markers/coarse_cluster_markers_all_data_soupx.tsv")

topx <- mkrs %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  slice(1:3)

so <- ScaleData(so, features = unique(topx$feature))

p <- DoHeatmap(so, 
               group.colors = discrete_palette_default,
        features = unique(topx$feature),
        group.by = "coarse_clusters",
        angle = 0,
        raster = FALSE, 
        draw.lines = TRUE,
        label = FALSE)

p <- p +
  scale_fill_gradientn(colours = viridis::viridis(256),
                       name = "Log-Normalized Counts\nZ-scores")

save_plot(file.path(all_cell_outdir, "all_data_marker_heatmap.pdf"), 
          p,
          base_asp = 1.75,
          base_height = 12)

p
```

#### Subset cells to same minimum #

```{r, fig.width=12, fig.height=8}
min_cells <- min(table(so@meta.data$coarse_clusters) )
set.seed(42)
to_keep <- get_metadata(so) %>% 
  group_by(coarse_clusters) %>% 
  sample_n(size =  min_cells, replace = FALSE) %>% 
  pull(cell)
so_sub <- subset(so, cells = to_keep)

p <- DoHeatmap(so_sub, 
               group.colors = discrete_palette_default,
        features = unique(topx$feature),
        group.by = "coarse_clusters",
        angle = 0,
        raster = FALSE, 
        draw.lines = TRUE,
        label = FALSE)

p <- p +
  scale_fill_gradientn(colours = viridis::viridis(256),
                       name = "Log-Normalized Counts\nZ-scores")

save_plot(file.path(all_cell_outdir, "all_data_marker_heatmap_cell_subset.pdf"), 
          p,
          base_asp = 1.33,
          base_height = 12)

p
```
#### Show average

```{r, fig.width=12, fig.height=8}
library(ComplexHeatmap)

max_disp <- 2.5

avg_expr <- AverageExpression(so, 
                              use.scale = TRUE, 
                              features = unique(topx$feature))$RNA %>% 
  as.matrix()

avg_expr[avg_expr > max_disp] <- max_disp
avg_expr[avg_expr < -max_disp] <- -max_disp

group_ids <- colnames(avg_expr)

annot_df <- data.frame(
  clusters = factor(group_ids, levels = as.character(sort(as.numeric(group_ids))))
)
ha <- HeatmapAnnotation(df = annot_df,
                        col = list(
                          clusters = structure(discrete_palette_default[1:length(group_ids)],
                                               names = group_ids)
                        ),
                        annotation_legend_param = list(
        clusters = list(ncol = 1))
)

hmat <- Heatmap(avg_expr[unique(topx$feature), ],
                name = "Log Normalized\nAverge Z-scores",
                col = viridis::viridis(256),
             cluster_rows = FALSE,
             cluster_columns = FALSE,
             row_names_side = "left",
             column_names_side = "top",
             column_names_rot = 0,
             top_annotation = ha)

draw(hmat, merge_legend = TRUE)


pdf(file.path(all_cell_outdir, "all_data_marker_heatmap_average.pdf"),
    height = 9,
    width = 9)
  draw(hmat, merge_legend = TRUE)
dev.off()
```

## Endothelial cells analysis

```{r}
endo_cell_outdir <- file.path(fig_dir, "endo_cells")
dir.create(endo_cell_outdir, showWarnings = FALSE)
```

```{r}
so <- qread(file.path("objects", "so_endo.qs"))
```


### UMAP plots {.tabset}


#### Cell type


```{r}
p <- plot_umap(so, 
          feature = "refined_clusters", 
          sorted = "random",
          pt_alpha = 1, 
          legend_title = "",
          label_text = TRUE)

p
save_plot(file.path(endo_cell_outdir,
                    "umap_by_cluster.pdf"),
          p,
          base_asp = 1.6)
```


#### Condition

```{r}
p <- plot_umap(so, 
          feature = "condition", 
          sorted = "random",
          .cols = palette_OkabeIto,
          pt_alpha = 0.6, 
          legend_title = "")

p
save_plot(file.path(endo_cell_outdir,
                    "umap_by_condition.pdf"),
          p,
          base_asp = 1.6)
```


#### Cell types + Condition

```{r}
p <- plot_umap(so, 
          feature = "refined_clusters", 
          sorted = "random",
          group = "condition",
          pt_alpha = 0.6, 
          legend_title = "")

p
save_plot(file.path(endo_cell_outdir,
                    "umap_by_condition_split.pdf"),
          p, ncol = 2,
          base_asp = 1.5)
```

#### Sample

```{r}
p <- plot_umap(so, 
          feature = "orig.ident", 
          sorted = "random",
          .cols = palette_OkabeIto,
          pt_alpha = 1, 
          legend_title = "")

p
save_plot(file.path(endo_cell_outdir,
                    "umap_by_sample.pdf"),
          p,
          base_asp = 1.6)
```

### UMAP by Gene

```{r, results ='asis'}

to_plot <- c(
  "Cldn5",
  "mt-Atp6",
  "Emp2",
  "Hbb-bt",
  "Vwf",
  "Cd3e",
  "Col1a2",
  "Pf4",
  "Vtn",
  "Cd79a",
  "Prox1")


dir.create(file.path(endo_cell_outdir, "umap_by_gene"), 
           showWarnings = FALSE)
res <- list()
for(i in seq_along(to_plot)){
  cat('\n#### ', to_plot[i], '\n')
  
  p <- plot_umap(so, to_plot[i])
  
  print(p)
  res[[i]] <- p
  save_plot(file.path(endo_cell_outdir,
                      "umap_by_gene",
                    paste0("umap_by_", to_plot[i], ".pdf")),
          p,
          base_asp = 1.3)
  cat('\n')
}
```

#### All markers

```{r, fig.width = 12, fig.height = 16}
p <- plot_grid(plotlist = res, 
          nrow = 3,
          ncol = 4)

p
save_plot(file.path(endo_cell_outdir,
                     "umap_all_selected_markers.pdf"),
          nrow = 3, ncol = 4,
          p,
          base_asp = 1.3)
 
```

#### All markers minimal theme

```{r, fig.width = 12, fig.height = 16}
p <- plot_umap(so, to_plot, minimal_theme = TRUE) %>% 
  plot_grid(plotlist = ., 
          nrow = 3,
          ncol = 4)

p
save_plot(file.path(endo_cell_outdir,
                     "umap_all_selected_markers_minimal_theme.pdf"),
          nrow = 3, ncol = 4,
          p,
          base_asp = 1.0)
 
```


### Violin plots {.tabset}

```{r, results ='asis'}
to_plot <- c(
  "Cldn5",
  "mt-Atp6",
  "Emp2",
  "Col1a2",
  "Hbb-bt",
  "Vwf",
  "Cd3e",
  "Pf4",
  "Prox1",
  "Vtn",
  "Cd79a"
)

dir.create(file.path(endo_cell_outdir, "vln_by_gene"), showWarnings = FALSE)
res <- list()
for(i in seq_along(to_plot)){
  cat('\n#### ', to_plot[i], '\n')
  
  p <- scbp::plot_violins(so, 
            group = "refined_clusters", 
            features = to_plot[i]) +
  labs(x = "") + 
  theme(legend.position = "none")
  print(p)
  
  res[[i]] <- p
  save_plot(file.path(endo_cell_outdir,
                      "vln_by_gene",
                    paste0("vln_by_", to_plot[i], ".pdf")),
          p,
          base_asp = 1.5)
  cat('\n')
}

```

#### All markers

```{r}
p <- scbp::plot_violins(so, 
            group = "refined_clusters", 
            features = to_plot) +
  labs(x = "",
       y = "") + 
  theme(legend.position = "none")
p
save_plot(file.path(endo_cell_outdir,
                     "vln_all_selected_markers.pdf"),
          p,
          base_asp = 1.0,
          base_height = 14)

```

#### All markers with condition

```{r}
p <- scbp::plot_violins(so, 
            group = "refined_clusters", 
            features = to_plot,
            split_by = "condition") +
  labs(x = "",
       y = "") 
p
save_plot(file.path(endo_cell_outdir,
                     "vln_all_selected_markers_by_condition.pdf"),
          p,
          base_asp = 1.0,
          base_height = 14)

```

### Cell Proportions {.tabset}

#### Condition by cell type

```{r}
p <- plot_cell_proportions(so, "refined_clusters", "condition") +
  labs(x = "",
       fill = "") 
  
p

save_plot(file.path(endo_cell_outdir, 
                    "cell_proportion_clusters_by_condition.pdf"),
          p,
          base_asp = 1.2,
          base_height = 5)
```


#### Condition by sample

```{r}
p <- plot_cell_proportions(so, "refined_clusters", "orig.ident") +
  labs(x = "",
       fill = "")
p

save_plot(file.path(endo_cell_outdir, 
                    "cell_proportion_clusters_by_sample.pdf"),
          p,
          base_asp = 1.2,
          base_height = 5)
```


### Heatmaps

```{r, fig.width=12, fig.height=8}
mkrs <- read_tsv("markers/endothelial_subcluster_markers.tsv")

topx <- mkrs %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  slice(1:5)

so <- ScaleData(so, features = unique(topx$feature))

p <- DoHeatmap(so, 
               group.colors = discrete_palette_default,
        features = unique(topx$feature),
        group.by = "refined_clusters",
        angle = 0,
        raster = FALSE, 
        draw.lines = TRUE,
        label = FALSE)

p <- p +
  scale_fill_gradientn(colours = viridis::viridis(256),
                       name = "Log-Normalized Counts\nZ-scores")

p

save_plot(file.path(endo_cell_outdir, "endo_marker_heatmap.pdf"), 
          p,
          base_asp = 2,
          base_height = 12)
```

## B-cell analysis

```{r}
b_cell_outdir <- file.path(fig_dir, "b_cells")
dir.create(b_cell_outdir, showWarnings = FALSE)
```

```{r}
so <- qread(file.path("objects", "bcell_subclustering.qs"))
```


### UMAP plots {.tabset}


#### Cell type


```{r}
p <- plot_umap(so, 
          feature = "refined_clusters", 
          sorted = "random",
          pt_alpha = 1, 
          legend_title = "",
          label_text = TRUE)

p
save_plot(file.path(b_cell_outdir,
                    "umap_by_cluster.pdf"),
          p,
          base_asp = 1.6)
```


#### Condition

```{r}
p <- plot_umap(so, 
          feature = "condition", 
          sorted = "random",
          .cols = palette_OkabeIto,
          pt_alpha = 0.6, 
          legend_title = "")

p
save_plot(file.path(b_cell_outdir,
                    "umap_by_condition.pdf"),
          p,
          base_asp = 1.6)
```

#### Cell types + Condition

```{r}
p <- plot_umap(so, 
          feature = "coarse_clusters", 
          sorted = "random",
          group = "condition",
          pt_alpha = 0.6, 
          legend_title = "")

p
save_plot(file.path(b_cell_outdir,
                    "umap_by_condition_split.pdf"),
          p, ncol = 2,
          base_asp = 1.5)
```

#### Sample

```{r}
p <- plot_umap(so, 
          feature = "orig.ident", 
          sorted = "random",
          .cols = palette_OkabeIto,
          pt_alpha = 1, 
          legend_title = "")

p
save_plot(file.path(b_cell_outdir,
                    "umap_by_sample.pdf"),
          p,
          base_asp = 1.6)
```

#### Pseudotime

```{r}
p <- plot_umap(so, 
          feature = "Pseudotime", 
          sorted = "random",
          pt_alpha = 1, 
          legend_title = "")

p
save_plot(file.path(b_cell_outdir,
                    "umap_by_pseudotime.pdf"),
          p,
          base_asp = 1.2)
```

#### CytoTRACE

```{r}
p <- plot_umap(so, 
          feature = "CytoTRACE", 
          sorted = "random",
          pt_alpha = 1, 
          legend_title = "")

p
save_plot(file.path(b_cell_outdir,
                    "umap_by_cytotrace.pdf"),
          p,
          base_asp = 1.2)
```

### UMAP by gene {.tabset}

```{r, results ='asis'}

to_plot <- c(
  "Mzb1",
  "Zbtb32",
  "Bhlhe41",
  "Itgb1",
  "Iglv1",
  "Jchain",
  "Prg2",
  "Cd79a",
  "Ighm",
  "Cxcr3")


dir.create(file.path(b_cell_outdir, "umap_by_gene"), 
           showWarnings = FALSE)
res <- list()
for(i in seq_along(to_plot)){
  cat('\n#### ', to_plot[i], '\n')
  
  p <- plot_umap(so, to_plot[i])
  
  print(p)
  res[[i]] <- p
  save_plot(file.path(b_cell_outdir,
                      "umap_by_gene",
                    paste0("umap_by_", to_plot[i], ".pdf")),
          p,
          base_asp = 1.3)
  cat('\n')
}
```

#### All markers

```{r, fig.width = 12, fig.height = 16}
p <- plot_grid(plotlist = res, 
          nrow = 3,
          ncol = 4)

p
save_plot(file.path(b_cell_outdir,
                     "umap_all_selected_markers.pdf"),
          nrow = 3, ncol = 4,
          p,
          base_asp = 1.3)
 
```

#### All markers minimal theme

```{r, fig.width = 12, fig.height = 16}
p <- plot_umap(so, to_plot, minimal_theme = TRUE) %>% 
  plot_grid(plotlist = ., 
          nrow = 3,
          ncol = 4)

p
save_plot(file.path(b_cell_outdir,
                     "umap_all_selected_markers_minimal_theme.pdf"),
          nrow = 3, ncol = 4,
          p,
          base_asp = 1.0)
 
```



### UMAP of antibody isotypes

```{r, results ='asis'}

to_plot <- str_subset(rownames(so), "Igh") %>%
  .[!str_detect(., "Ighv[1-9]")]

dir.create(file.path(b_cell_outdir, "umap_by_isotype"), 
           showWarnings = FALSE)
res <- list()
for(i in seq_along(to_plot)){
  cat('\n#### ', to_plot[i], '\n')
  
  p <- plot_umap(so, to_plot[i])
  
  print(p)
  res[[i]] <- p
  save_plot(file.path(b_cell_outdir,
                      "umap_by_isotype",
                    paste0("umap_by_", to_plot[i], ".pdf")),
          p,
          base_asp = 1.3)
  cat('\n')
}
```

#### All markers

```{r, fig.width = 12, fig.height = 16}
p <- plot_grid(plotlist = res, 
          nrow = 4,
          ncol = 2)

p
save_plot(file.path(b_cell_outdir,
                     "umap_all_isotypes.pdf"),
          nrow = 4, ncol = 2,
          p,
          base_asp = 1.3)
 
```

#### All markers minimal theme

```{r, fig.width = 12, fig.height = 16}
p <- plot_umap(so, to_plot, minimal_theme = TRUE) %>% 
  plot_grid(plotlist = ., 
          nrow = 4,
          ncol = 2)

p
save_plot(file.path(b_cell_outdir,
                     "umap_all_isotypes_minimal_theme.pdf"),
          nrow = 4, ncol = 2,
          p,
          base_asp = 1.0)
 
```



### Violin plots {.tabset}

```{r, results ='asis'}
to_plot <- str_subset(rownames(so), "Igh") %>%
  .[!str_detect(., "Ighv[1-9]")]

dir.create(file.path(b_cell_outdir, "vln_by_gene"), showWarnings = FALSE)
res <- list()
for(i in seq_along(to_plot)){
  cat('\n#### ', to_plot[i], '\n')
  
  p <- scbp::plot_violins(so, 
            group = "coarse_clusters", 
            features = to_plot[i]) +
  labs(x = "") + 
  theme(legend.position = "none")
  print(p)
  
  res[[i]] <- p
  save_plot(file.path(b_cell_outdir,
                      "vln_by_gene",
                    paste0("vln_by_", to_plot[i], ".pdf")),
          p,
          base_asp = 1.5)
  cat('\n')
}

```

#### All markers

```{r}
p <- scbp::plot_violins(so, 
            group = "coarse_clusters", 
            features = to_plot) +
  labs(x = "",
       y = "") + 
  theme(legend.position = "none")
p
save_plot(file.path(b_cell_outdir,
                     "vln_all_selected_markers.pdf"),
          p,
          base_asp = 1.0,
          base_height = 14)

```

#### All markers with condition

```{r}
p <- scbp::plot_violins(so, 
            group = "coarse_clusters", 
            features = to_plot,
            split_by = "condition") +
  labs(x = "",
       y = "") 
p
save_plot(file.path(b_cell_outdir,
                     "vln_all_selected_markers_by_condition.pdf"),
          p,
          base_asp = 1.0,
          base_height = 14)

```

### Cell Proportions {.tabset}

#### Condition by cell type

```{r}
p <- plot_cell_proportions(so, "refined_clusters", "condition") +
  labs(x = "",
       fill = "") 
  
p

save_plot(file.path(b_cell_outdir, 
                    "cell_proportion_refined_clusters_by_condition.pdf"),
          p,
          base_asp = 1.2,
          base_height = 5)
```

```{r}
p <- plot_cell_proportions(so, "coarse_clusters", "condition") +
  labs(x = "",
       fill = "") 
  
p

save_plot(file.path(b_cell_outdir, 
                    "cell_proportion_coarse_clusters_by_condition.pdf"),
          p,
          base_asp = 1.2,
          base_height = 5)
```

#### Condition by sample

```{r}
p <- plot_cell_proportions(so, "refined_clusters", "orig.ident") +
  labs(x = "",
       fill = "")
p

save_plot(file.path(b_cell_outdir, 
                    "cell_proportion_refined_clusters_by_sample.pdf"),
          p,
          base_asp = 1.2,
          base_height = 5)
```


```{r}
p <- plot_cell_proportions(so, "coarse_clusters", "orig.ident") +
  labs(x = "",
       fill = "")
p

save_plot(file.path(b_cell_outdir, 
                    "cell_proportion_coarse_clusters_by_sample.pdf"),
          p,
          base_asp = 1.2,
          base_height = 5)
```

#### Better represent samples

```{r}
x <- get_cell_count_matrix(so, "coarse_clusters", "orig.ident") 
x <-  t(t(x) / colSums(x)) 
plt_dat <- x %>% 
  as.data.frame() %>% 
  rownames_to_column("cell_type") %>% 
  pivot_longer(cols = -cell_type,
               names_to = "sample",
               values_to = "count") %>% 
  mutate(count = 100 * count) %>% 
  separate(sample, c("genotype", "rep"), sep = "_") 

p <- ggplot(plt_dat, aes(genotype, count)) +
  geom_boxplot(aes(fill =genotype), outlier.shape = NA) +
  geom_jitter(aes(group = genotype)) +
  facet_wrap(~cell_type, scales = "free", strip.position = "top", nrow = 1) +
  scale_fill_brewer(palette = "Paired") + 
  labs(x = "",
       y = "% of cells") + 
  theme(strip.placement = "outside",
        legend.position = "none")
p
save_plot(file.path(b_cell_outdir, 
                    "cell_proportion_coarse_clusters_by_condition_all_samples.pdf"),
          p,
          base_asp = 3.0)

cell_type_stats <- calc_da(so, "coarse_clusters", "orig.ident", "condition")

write_tsv(cell_type_stats, file.path(all_cell_outdir, 
                    "cell_proportion_coarse_clusters_by_condition_all_samples_stats.tsv"))
```

### Selected markers

Rerun markers compareing 1 and 2 versus other clusters

```{r}
library(presto)
so_tmp <- subset(so, subset = coarse_clusters %in% c("0", "1", "2"))
so_tmp$tmp_groups <- ifelse(so_tmp$coarse_clusters %in% c("1", "2"),
                        "mzb1-like-cells",
                        "other-b-cells")
mzb1_markers <- wilcoxauc(so_tmp, "tmp_groups") %>% 
  filter(pct_in > 5, padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE)

scbp::write_markers_xlsx(list(cluster_1_2_v_0 = mzb1_markers), 
                   file.path(tbls_dir,
                             "b_cell_subclustered_cluster_1and2_v_others.xlsx"),
                   description_string = "Differentially expressed genes between cluster 1 + 2 against clusters 0). Position logFC indicates upregulated in clusters 1 and 2. Genes were filtered for padj < 0.05, and requiring at least 5% of the cells in clusters 1 and 2 to express the gene"
                   )
```

### Heatmaps

#### All data
```{r, fig.width=12, fig.height=8}
mkrs <- read_tsv("markers/bcells_mzb1_markers_coarse_clusters.tsv")

topx <- mkrs %>% 
  filter(padj < 1e-4) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  slice(1:10) %>% 
  split(., .$group) %>% 
  map(., ~pull(.x, feature))

# add in mzb1 for clarity
topx$`2` <- c(topx$`2`, "Mzb1")
topx <- unlist(topx)
so <- ScaleData(so, features = unique(topx))

p <- DoHeatmap(so, 
               group.colors = discrete_palette_default,
        features = topx,
        group.by = "coarse_clusters",
        angle = 0,
        raster = FALSE, 
        draw.lines = TRUE,
        label = FALSE)

p <- p +
  scale_fill_gradientn(colours = viridis::viridis(256),
                       name = "Log-Normalized Counts\nZ-scores")

p

save_plot(file.path(b_cell_outdir, "bcell_marker_heatmap.pdf"), 
          p,
          base_asp = 1.5,
          base_height = 12)
```


#### Subset cells to same minimum #

```{r, fig.width=12, fig.height=8}
min_cells <- min(table(so@meta.data$coarse_clusters) )
set.seed(42)
to_keep <- get_metadata(so) %>% 
  group_by(coarse_clusters) %>% 
  sample_n(size =  min_cells, replace = FALSE) %>% 
  pull(cell)
so_sub <- subset(so, cells = to_keep)

p <- DoHeatmap(so_sub, 
               group.colors = discrete_palette_default,
        features = unique(topx),
        group.by = "coarse_clusters",
        angle = 0,
        raster = FALSE, 
        draw.lines = TRUE,
        label = FALSE)

p <- p +
  scale_fill_gradientn(colours = viridis::viridis(256),
                       name = "Log-Normalized Counts\nZ-scores")

save_plot(file.path(b_cell_outdir, "bcell_marker_heatmap_cell_subset.pdf"), 
          p,
          base_asp = 1.33,
          base_height = 12)

p
```

#### Show average

```{r, fig.width=12, fig.height=8}
library(ComplexHeatmap)

max_disp <- 2.5
Idents(so) <- "coarse_clusters"
avg_expr <- AverageExpression(so, 
                              slot= "scale.data", 
                              features = unique(topx))$RNA %>% 
  as.matrix()

avg_expr[avg_expr > max_disp] <- max_disp
avg_expr[avg_expr < -max_disp] <- -max_disp

group_ids <- colnames(avg_expr)

annot_df <- data.frame(
  clusters = factor(group_ids, levels = as.character(sort(as.numeric(group_ids))))
)
ha <- HeatmapAnnotation(df = annot_df,
                        col = list(
                          clusters = structure(discrete_palette_default[1:length(group_ids)],
                                               names = group_ids)
                        ),
                        annotation_legend_param = list(
        clusters = list(ncol = 1))
)

hmat <- Heatmap(avg_expr[unique(topx), ],
                name = "Log Normalized\nAverge Z-scores",
                col = viridis::viridis(256),
             cluster_rows = FALSE,
             cluster_columns = FALSE,
             row_names_side = "left",
             column_names_side = "top",
             column_names_rot = 0,
             top_annotation = ha)

draw(hmat, merge_legend = TRUE)


pdf(file.path(b_cell_outdir, "bcell_marker_heatmap_average.pdf"),
    height = 9,
    width = 9)
  draw(hmat, merge_legend = TRUE)
dev.off()
```

#### just show clusters 1 and 2 (the most interesting)

```{r, fig.width=12, fig.height=8}

mkrs <- read_tsv("markers/bcells_mzb1_markers_coarse_clusters.tsv")

topx <- mkrs %>% 
  filter(padj < 1e-4,
         group %in% c(1, 2)) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  slice(1:10) %>% 
  split(., .$group) %>% 
  map(., ~pull(.x, feature))

topx$`2` <- c(topx$`2`, "Mzb1")
topx <- unlist(topx)

so <- ScaleData(so, features = rownames(so))
so_sub <- subset(so, cells = to_keep)

p <- DoHeatmap(so_sub, 
               group.colors = discrete_palette_default,
        features = unique(topx),slot = "scale.data",
        group.by = "coarse_clusters",
        angle = 0,
        raster = FALSE, 
        draw.lines = TRUE,
        label = FALSE)

p <- p +
  scale_fill_gradientn(colours = viridis::viridis(256),
                       name = "Log-Normalized Counts\nZ-scores")

save_plot(file.path(b_cell_outdir, "bcell_marker_heatmap_cluster_subset.pdf"), 
          p,
          base_asp = 1.0,
          base_height = 7)

p
```


#### clusters 1 and 2 vs 0 

```{r, fig.width=12, fig.height=8}
topn_vals <- 40
top_up <- mzb1_markers %>% 
  filter(padj < 1e-4,
         group == "mzb1-like-cells",
         logFC > 0) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  slice(1:topn_vals) %>%
  pull(feature)

top_down <- mzb1_markers %>% 
  filter(padj < 1e-4,
         group == "mzb1-like-cells",
         logFC < 0) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  slice(1:topn_vals) %>%
  pull(feature)
topx <- c(top_up, top_down)

so_tmp <- ScaleData(so_tmp, features = rownames(so_tmp))
so_sub <- subset(so_tmp, cells = to_keep)

so_sub$coarse_clusters <- droplevels(so_sub$coarse_clusters)
# p <- DoHeatmap(so_sub, 
#                group.colors = discrete_palette_default,
#         features = unique(topx),slot = "scale.data",
#         group.by = "coarse_clusters",
#         angle = 0,
#         raster = FALSE, 
#         draw.lines = TRUE,
#         label = FALSE)
# 
# p <- p +
#   scale_fill_gradientn(colours = viridis::viridis(256),
#                        name = "Log-Normalized Counts\nZ-scores")


hmap <- plot_heatmap(so_sub, 
             features = topx, 
             group = "coarse_clusters", 
             hmap_options = list(cluster_rows = TRUE))

pdf(file.path(b_cell_outdir, 
              "bcell_marker_heatmap_cluster_1_and_2 _markers_subset.pdf"),
    width = 12 * 1.2,
    height = 12)
ComplexHeatmap::draw(hmap)
dev.off()
# save_plot(file.path(b_cell_outdir, "bcell_marker_heatmap_cluster_1_and_2 _markers_subset.pdf"), 
#           p,
#           base_asp = 1.2,
#           base_height = 12)
# 
# p
```
